<a href="https://colab.research.google.com/github/aka-hombre/MAT421/blob/main/HW_CH9.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [2]:
import numpy

#**CHAPTER 9**
##**Representing Base n and binary**

The numbers we are familiar with are decimal, base-10.

This means our numbers can be represented as a sum of powers, with a base of ten.

Example: $111 = 1*10^{2} + 1*10^{1} + 1*10^{0}$

Now we can do the same for a diffrent base, take base 7 for example. To convert we can divide our decimal number the remainder will be the digits of our number. We will use some modular arithimitic to find the coefficients.

In [18]:
def to_base_n(number, base): #function to find coeff. of a dec number to base n
  coeff = []
  while(number/base != 0):
    coeff.append(number%base) #finding the remainders of the number
    number = number//base #performing floor division
  return coeff


print(to_base_n(111,7))

[6, 1, 2]


$111 = 2*7^{2} +1*7^{1}+6*7^{0}$

So 111 in base 7 is 216!

Binary is the same thing, but we work in base 2. Menaing, our remainders from division are always going to be 0 or 1.

Let's check with our function to find the coefficients

In [19]:
print(to_base_n(111,2))

[1, 1, 1, 1, 0, 1, 1]


$111 = 1 * 2^6+1*2^5+0*2^4+1*2^3+1*2^2+1*2^1+1*2^0$

So 111 in binary is:  1101111

Python can also do binary conversions on it's own too.

In [17]:
print("111 in binary: "+bin(111))

111 in binary: 0b1101111



Here are some example of arithmetic of binary numbers

In [27]:
a = 55
b = 72
c = 33

print("a to binary:"+bin(a)+"\n"+"b to binary:"+bin(b)+"\n"+"c to binary:"+bin(c)+"\n")

x = bin(a)+bin(b)
y = bin(a*b)
z = bin(c*b)

print("0b110111+0b1001000="+x)
print("0b100001*0b1001000="+z)
print("0b1001000*0b110111="+y)

a to binary:0b110111
b to binary:0b1001000
c to binary:0b100001

0b110111+0b1001000=0b1101110b1001000
0b100001*0b1001000=0b100101001000
0b1001000*0b110111=0b111101111000


##**Floating Point numbers**

In [28]:
import sys
sys.float_info

sys.float_info(max=1.7976931348623157e+308, max_exp=1024, max_10_exp=308, min=2.2250738585072014e-308, min_exp=-1021, min_10_exp=-307, dig=15, mant_dig=53, epsilon=2.220446049250313e-16, radix=2, rounds=1)

Floating point numbers need to be represented in three parts: sign, exponent, and fraction. Each part has bits allocated to it.

This means there is are numbers not in the bounds of memory on a computer, so certain numbers will either *overflow* or *underflow*

In [62]:
maxnum = 1.7976931348623157*(10**308)
minnum = 2.2250738585072014*(10**-308)
print("Max supported number: "+ str(maxnum))
print("Adding one to the max number won't register: "+str(1+maxnum))
print("Multiplying by two, will overflow it: "+str(2*maxnum))
print("\n")
print("Minimum supported number: "+ str(minnum))
print("Integer division by two, will overflow it: "+str(minnum//2))

Max supported number: 1.7976931348623157e+308
Adding one to the max number won't register: 1.7976931348623157e+308
Multiplying by two, will overflow it: inf


Minimum supported number: 2.2250738585072014e-308
Integer division by two, will overflow it: 0.0


This also means there are gaps between the floating point number.



In [70]:
print("Here is the spacing between the number 111.3: "+str(numpy.spacing(111.3)))

print("Adding 1/3 of the space to 111.3 yeilds: "+str(111.3 + (numpy.spacing(111.3)/3)))

Here is the spacing between the number 111.3: 1.4210854715202004e-14
Adding 1/3 of the space to 111.3 yeilds: 111.3


##**Round off errors**

These occur whenthe numerical mehtod of calculation differs from the arithmetic computed by a computer. This happens because floating points cannot be represented exactly by a number


In [71]:
print(3.14-3.00005)

0.13995000000000024


The numerical method yeilds: $3.14-3.00005=0.13995$


As we can see this is not the value python gives

We can use the round function to get better results.

In [72]:
round(3.14-3.00005,5)

0.13995

Round off errors can accumulate we will add 1/3 and then subtract 1/3 from three, a fixed amount of times to see the error acumulated with each itteration. The actual value should be 3, as we know.

In [75]:
def add_and_subtract(iterations):
    result = 3

    for i in range(iterations):
        result += 2/3

    for i in range(iterations):
        result -= 2/3
    return result

print("10 times itterated: "+str(add_and_subtract(10)))
print("100 times itterated: "+str(add_and_subtract(100)))
print("1000 times itterated: "+str(add_and_subtract(1000)))

10 times itterated: 2.999999999999999
100 times itterated: 3.000000000000001
1000 times itterated: 3.000000000000015
